### PR TITLE
Move todo comments analyzer to Editor Features

### PR DESCRIPTION
**Pull request by davidwengier**
FYI @Therzok 

There isn't much to this so I don't think its going to fundamentally fix anything in VS Mac, but we may as well prevent them having to copy the source code if there isn't a good reason to.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
> Move todo comments analyzer to Editor Features 

Can this not be at the feature level?

**Comment by davidwengier** [comment]
> > Move todo comments analyzer to Editor Features
> 
> Can this not be at the Features or even Workspace level? I don't see anything Editor specific about this (i.e. using buffers/views).

Updated. Cunningham's Law strikes again :)

**Comment by sharwell** [comment]
@davidwengier Don't merge this yet. It was carefully constructed to silently disable itself if OOP gets turned off. I want to make sure that 1) we don't lose that and 2) we aren't running this analyzer in both processes at the same time.

File: src/EditorFeatures/CSharpTest/TodoComment/NoCompilationTodoCommentTests.cs
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/TodoComment/NoCompilationTodoCommentTests.cs:60]: ❔ Can this not be fixed with a using directive?

**Review by sharwell — CHANGES_REQUESTED** [review]
I have no strong objection to trying to move `InProcTodoCommentsIncrementalAnalyzer` down to the Features layer, but there is a big remaining problem: `InProcTodoCommentsIncrementalAnalyzerProvider` is not exported (claimed by design), which means this move does not make TODO comments scanning any easier for external consumers. We need to either move `VisualStudioTodoCommentsService` or find a way to register the provider without breaking functionality.

**Comment by davidwengier** [comment]
@sharwell It's not clear to me why we need to do that before merging this PR, given that it wasn't a goal of this PR nor is made any more difficult by this PR.

**Comment by allisonchou** [comment]
/azp run
(Going through open PRs since it looks like the master -> main transition made test running funky.)

**Comment by sharwell** [comment]
@davidwengier It's not clear what the purpose of this pull request is. `AbstractTodoCommentsIncrementalAnalyzer` already lives in the features layer, and the changes proposed by this pull request do not align with the claim that it removes the need to implement these types separately (as the types are not exposed).

**Comment by davidwengier** [comment]
> @davidwengier David Wengier FTE It's not clear what the purpose of this pull request is.

The purpose is solely to prevent VSMac having to source-copy the incremental analyzer and analyzer provider. It comes from a request [here](https://github.com/xamarin/vsmac/pull/2495#discussion_r511564161) if you have access.

> it's not desirable to adjust the code in ways that specifically facilitate increased use of APIs through IVTs.

I agree with the spirit of this comment, but when the alternative option is copying source code which then becomes stale, I think allowing the IVT is the objectively better choice. I think thats especially true for something like this which is a pure move. If I'd left the namespaces in tact in the files there would have been no code changes at all.

For good or bad VSMac has extensive use of IVTs and I don't feel that blocking individual PRs does anything other than make life harder for the VSMac team, or in some cases prevents them shipping software. There was a larger conversation around trying to move them to external access but it was ruled out because the scope is too large, and the resources too little.

**Comment by sharwell** [comment]
In this case it seems desirable to keep the current form. When we do move the implementation, it should be moved and exposed only through an external access layer which VS Mac can use.

> There was a larger conversation around trying to move them to external access but it was ruled out because the scope is too large, and the resources too little.

This was the case when VS Mac agreed to handle breaking changes on their side as they arose, but it is proving to not be viable. VS Mac will need to transition to external access and progressively revoke the external IVTs so these problems do not continue in the future.

**Comment by davidwengier** [comment]
Desirable to who? The VS Mac team have asked for it (that comment I linked to is an open PR on their end, refactoring their todo comment implementation) and yes this will break VS Mac, but i'm the one who fixes those breaks, so if you're blocking my PR in order to save me work, you may rest assured that I considered that in advance 😛

**Review by jasonmalinowski — APPROVED** [review]

**Pull request merged**
